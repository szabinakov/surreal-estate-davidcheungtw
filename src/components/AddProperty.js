import React, { useState } from "react";
import axios from "axios";
import Alert from "./Alert";
import "../styles/add-property.css";

const AddProperty = () => {
  const cities = ["Manchester", "Leeds", "Sheffield", "Liverpool"];
  const types = [
    "Flat",
    "Detached",
    "Semi-Detached",
    "Terraced",
    "End of Terrace",
    "Cottage",
    "Bungalow",
  ];
  const initialState = {
    fields: {
      title: "",
      city: "Manchester",
      type: "Flat",
      bedrooms: 0,
      bathrooms: 0,
      price: 0,
      email: "",
    },
    alert: {
      message: "",
      isSuccess: false,
    },
  };
  const [fields, setFields] = useState(initialState.fields);
  const [alert, setAlert] = useState(initialState.alert);

  const handleFieldChange = (event) => {
    setFields({ ...fields, [event.target.name]: event.target.value });
  };

  const handleAddProperty = async (event) => {
    event.preventDefault();
    const { title, city, type, bedrooms, bathrooms, price, email } = fields;

    setAlert({
      message: "",
      isSuccess: false,
    });

    if (
      title &&
      city &&
      type &&
      bedrooms > 0 &&
      bathrooms > 0 &&
      price > 0 &&
      email
    ) {
      axios
        .post("http://localhost:4000/api/v1/PropertyListing", fields)
        .then(() => {
          setAlert({
            message: "Property added.",
            isSuccess: true,
          });
        })
        .catch(() => {
          setAlert({
            message: "Server error. Please try again later.",
            isSuccess: false,
          });
        });

      return;
    }
    setAlert({
      message: "Please enter missing information!",
      isSuccess: false,
    });
  };

  return (
    <div className="add-property">
      <form className="add-property-form" onSubmit={handleAddProperty}>
        <Alert message={alert.message} isSuccess={alert.isSuccess} />
        <label htmlFor="title">
          Property Description
          <input
            type="text"
            id="title"
            name="title"
            value={fields.title}
            onChange={handleFieldChange}
            placeholder="2 bed flat"
            className="input-text"
          />
        </label>
        <label htmlFor="type">
          Type
          <select
            id="type"
            name="type"
            value={fields.type}
            onChange={handleFieldChange}
            className="select-type"
          >
            {types.map((type) => (
              <option key={type} value={type}>
                {type}
              </option>
            ))}
          </select>
        </label>
        <label htmlFor="bedrooms">
          Bedrooms
          <input
            type="number"
            id="bedrooms"
            name="bedrooms"
            value={fields.bedrooms}
            onChange={handleFieldChange}
            placeholder="0"
            step="1"
            min="0"
            max="9"
            className="input-number"
          />
        </label>
        <label htmlFor="bathrooms">
          Bathrooms
          <input
            type="number"
            id="bathrooms"
            name="bathrooms"
            value={fields.bathrooms}
            onChange={handleFieldChange}
            placeholder="0"
            step="1"
            min="0"
            max="9"
            className="input-number"
          />
        </label>
        <label htmlFor="price">
          Price
          <input
            type="number"
            id="price"
            name="price"
            value={fields.price}
            onChange={handleFieldChange}
            placeholder="0"
            step="100"
            min="0"
            className="input-number"
          />
        </label>
        <label htmlFor="city">
          City
          <select
            id="city"
            name="city"
            value={fields.city}
            onChange={handleFieldChange}
            className="select-city"
          >
            {cities.map((city) => (
              <option key={city} value={city}>
                {city}
              </option>
            ))}
          </select>
        </label>
        <label htmlFor="email">
          Email
          <input
            type="email"
            id="email"
            name="email"
            value={fields.email}
            onChange={handleFieldChange}
            placeholder="john.smith@email.co.uk"
            className="input-email"
          />
        </label>
        <button type="submit" className="button-add">
          Add
        </button>
      </form>
    </div>
  );
};

export default AddProperty;
