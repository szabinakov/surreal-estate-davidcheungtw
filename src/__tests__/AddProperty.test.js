import React from "react";
import { fireEvent, render, screen, waitFor } from "@testing-library/react";
import renderer from "react-test-renderer";
import axios from "axios";
import AddProperty from "../components/AddProperty";

describe("AddProperty", () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  it("Renders as expected", () => {
    const rendered = renderer.create(<AddProperty />);

    expect(rendered).toMatchSnapshot();
  });

  it("Assert all labels are present", () => {
    render(<AddProperty />);

    expect(screen.getByText(/Property Description/i)).toBeInstanceOf(
      HTMLLabelElement
    );
    expect(screen.getByText(/Type/i)).toBeInstanceOf(HTMLLabelElement);
    expect(screen.getByText(/Bedrooms/i)).toBeInstanceOf(HTMLLabelElement);
    expect(screen.getByText(/Bathrooms/i)).toBeInstanceOf(HTMLLabelElement);
    expect(screen.getByText(/Price/i)).toBeInstanceOf(HTMLLabelElement);
    expect(screen.getByText(/City/i)).toBeInstanceOf(HTMLLabelElement);
    expect(screen.getByText(/Email/i)).toBeInstanceOf(HTMLLabelElement);
  });

  it("Assert correct number of inputs are present", () => {
    render(<AddProperty />);

    expect(screen.getAllByRole("textbox")).toHaveLength(2);
    expect(screen.getAllByRole("combobox")).toHaveLength(2);
    expect(screen.getAllByRole("spinbutton")).toHaveLength(3);
  });

  it("Assert a single button is present", () => {
    render(<AddProperty />);
    const buttons = screen.getAllByRole("button");

    expect(buttons).toHaveLength(1);
    expect(buttons[0]).toHaveTextContent("Add");
  });

  it("Renders a success post request to Surreal Estate API", async () => {
    jest.spyOn(axios, "post").mockReturnValueOnce(Promise.resolve());
    render(<AddProperty />);
    const textboxs = screen.getAllByRole("textbox");
    const comboboxs = screen.getAllByRole("combobox");
    const spinbutton = screen.getAllByRole("spinbutton");
    const button = screen.getByRole("button");

    fireEvent.change(textboxs[0], { target: { value: "2 bed flat" } });
    fireEvent.change(textboxs[1], {
      target: { value: "john.smith@email.co.uk" },
    });
    fireEvent.change(comboboxs[0], { target: { value: "Flat" } });
    fireEvent.change(comboboxs[1], { target: { value: "Manchester" } });
    fireEvent.change(spinbutton[0], { target: { value: 2 } });
    fireEvent.change(spinbutton[1], { target: { value: 2 } });
    fireEvent.change(spinbutton[2], { target: { value: 1230 } });
    fireEvent.click(button);

    await waitFor(() => {
      expect(screen.getByText("Property added.")).toBeInTheDocument();
    });
  });

  it("Renders a fail post request to Surreal Estate API", async () => {
    jest.spyOn(axios, "post").mockReturnValueOnce(Promise.reject());
    render(<AddProperty />);
    const textboxs = screen.getAllByRole("textbox");
    const comboboxs = screen.getAllByRole("combobox");
    const spinbutton = screen.getAllByRole("spinbutton");
    const button = screen.getByRole("button");

    fireEvent.change(textboxs[0], { target: { value: "2 bed flat" } });
    fireEvent.change(textboxs[1], {
      target: { value: "john.smith@email.co.uk" },
    });
    fireEvent.change(comboboxs[0], { target: { value: "Flat" } });
    fireEvent.change(comboboxs[1], { target: { value: "Manchester" } });
    fireEvent.change(spinbutton[0], { target: { value: 2 } });
    fireEvent.change(spinbutton[1], { target: { value: 2 } });
    fireEvent.change(spinbutton[2], { target: { value: 1230 } });
    fireEvent.click(button);

    await waitFor(() => {
      expect(
        screen.getByText("Server error. Please try again later.")
      ).toBeInTheDocument();
    });
  });
});
