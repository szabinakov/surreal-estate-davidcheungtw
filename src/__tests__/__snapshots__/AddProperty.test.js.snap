// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddProperty Renders as expected 1`] = `
<div
  className="add-property"
>
  <form
    className="add-property-form"
    onSubmit={[Function]}
  >
    <label
      htmlFor="title"
    >
      Property Description
      <input
        className="input-text"
        id="title"
        name="title"
        onChange={[Function]}
        placeholder="2 bed flat"
        type="text"
        value=""
      />
    </label>
    <label
      htmlFor="type"
    >
      Type
      <select
        className="select-type"
        id="type"
        name="type"
        onChange={[Function]}
        value="Flat"
      >
        <option
          value="Flat"
        >
          Flat
        </option>
        <option
          value="Detached"
        >
          Detached
        </option>
        <option
          value="Semi-Detached"
        >
          Semi-Detached
        </option>
        <option
          value="Terraced"
        >
          Terraced
        </option>
        <option
          value="End of Terrace"
        >
          End of Terrace
        </option>
        <option
          value="Cottage"
        >
          Cottage
        </option>
        <option
          value="Bungalow"
        >
          Bungalow
        </option>
      </select>
    </label>
    <label
      htmlFor="bedrooms"
    >
      Bedrooms
      <input
        className="input-number"
        id="bedrooms"
        max="9"
        min="0"
        name="bedrooms"
        onChange={[Function]}
        placeholder="0"
        step="1"
        type="number"
        value={0}
      />
    </label>
    <label
      htmlFor="bathrooms"
    >
      Bathrooms
      <input
        className="input-number"
        id="bathrooms"
        max="9"
        min="0"
        name="bathrooms"
        onChange={[Function]}
        placeholder="0"
        step="1"
        type="number"
        value={0}
      />
    </label>
    <label
      htmlFor="price"
    >
      Price
      <input
        className="input-number"
        id="price"
        min="0"
        name="price"
        onChange={[Function]}
        placeholder="0"
        step="100"
        type="number"
        value={0}
      />
    </label>
    <label
      htmlFor="city"
    >
      City
      <select
        className="select-city"
        id="city"
        name="city"
        onChange={[Function]}
        value="Manchester"
      >
        <option
          value="Manchester"
        >
          Manchester
        </option>
        <option
          value="Leeds"
        >
          Leeds
        </option>
        <option
          value="Sheffield"
        >
          Sheffield
        </option>
        <option
          value="Liverpool"
        >
          Liverpool
        </option>
      </select>
    </label>
    <label
      htmlFor="email"
    >
      Email
      <input
        className="input-email"
        id="email"
        name="email"
        onChange={[Function]}
        placeholder="john.smith@email.co.uk"
        type="email"
        value=""
      />
    </label>
    <button
      className="button-add"
      type="submit"
    >
      Add
    </button>
  </form>
</div>
`;
